### PR TITLE
preValidation hook

### PR DESCRIPTION
**Pull request by delvedor**
Hi everyone!
With this PR we are introducing a new hook: `preValidation`, you can see the [`Lifecycle`](https://github.com/fastify/fastify/blob/new-hooks/docs/Lifecycle.md) to understand where we are using it.

As discussed in https://github.com/fastify/fastify/pull/955 we are also adding the support for the route level `preValidation` and I've renamed `beforeHandler` in `preHandler` (should we add a deprecation warning?).

I did a small cleanup in `fastify.js`, by renaming `app` in `avvio` and merged `beforeHandler.test.js` in `router-options.test.js`.

Feedbacks (also for the name of the new hook) are more than welcome!

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by allevo** [comment]
Awesome! I'm 👍  for `preValidation `

**Comment by delvedor** [comment]
@mcollina nope :)
But a double check would be good!

**Comment by mcollina** [comment]
Is this `semver-major`, or are you targeting `next` for convenience?

File: fastify.js [hunk 1]
```diff
@@ -585,13 +595,26 @@ function build (options) {
...
-          opts.beforeHandler.forEach((h, i) => {
-            opts.beforeHandler[i] = h.bind(_fastify)
+      if (opts.preHandler == null && opts.beforeHandler != null) {
+        opts.preHandler = opts.beforeHandler
+      }
+      if (opts.preHandler) {
+        if (Array.isArray(opts.preHandler)) {
+          opts.preHandler.forEach((h, i) => {
+            opts.preHandler[i] = h.bind(_fastify)
+          })
```
Comment (mcollina) [inline fastify.js]: Can you refactor this into a `.map()`?

File: fastify.js [hunk 2]
```diff
@@ -585,13 +595,26 @@ function build (options) {
...
+        } else {
+          opts.preHandler = opts.preHandler.bind(_fastify)
+        }
+      }
+
+      if (opts.preValidation) {
+        if (Array.isArray(opts.preValidation)) {
+          opts.preValidation.forEach((h, i) => {
+            opts.preValidation[i] = h.bind(_fastify)
           })
```
Comment (mcollina) [inline fastify.js]: this would be fantastic with a `.map()`

File: lib/handleRequest.js [hunk 3]
```diff
@@ -6,7 +6,13 @@ const validateSchema = validation.validate
 const { hookRunner, hookIterator } = require('./hooks')
 const wrapThenable = require('./wrapThenable')
 
-function handleRequest (request, reply) {
+function handleRequest (err, request, reply) {
+  if (reply.sent === true) return
+  if (err) {
```
Comment (mcollina) [inline lib/handleRequest.js]: Can you add a comment why are we receiving an error here? can you make the check explicit `err !== null && err !== undefined`?

**Comment by mcollina** [comment]
I think we should add a deprecation warning using https://nodejs.org/api/process.html#process_process_emitwarning_warning_type_code_ctor (note that we have to use this form, as the other one is not available in Node 6).

File: fastify.js [hunk 4]
```diff
@@ -835,15 +860,17 @@ function build (options) {
       this[kLogLevel]
     )
 
-    app.once('preReady', () => {
+    avvio.once('preReady', () => {
       const context = this[kFourOhFourContext]
 
       const onRequest = this[kHooks].onRequest
-      const preHandler = this[kHooks].preHandler.concat(opts.beforeHandler || [])
+      const preValidation = this[kHooks].preValidation.concat(opts.preValidation || [])
```
Comment (nwoltman) [inline fastify.js]: These `preValidation` hooks aren't being bound to the fastify instance, but the ones for normal routes are. Is this intended?

File: fastify.js [hunk 5]
```diff
@@ -835,15 +860,17 @@ function build (options) {
...
+      const preHandler = this[kHooks].preHandler.concat(opts.beforeHandler || opts.preHandler || [])
```
Comment (nwoltman) [inline fastify.js]: The `opts.beforeHandler` hooks get bound to the fastify instance here: https://github.com/fastify/fastify/blob/4e047a873ebf4cd2eb5bcece57425704e62d35cc/fastify.js#L824-L832

However, the `opts.preHandler` hooks aren't bound. It looks like the code above should be made the same as how you [handled](https://github.com/fastify/fastify/pull/1153/files#diff-20da0b6140a4768d182af0d9b185ac28R598) `before/preHandler` hooks for normal routes.

**Comment by nwoltman** [comment]
The [`setNotFoundHandler` docs](https://github.com/fastify/fastify/blob/4e047a873ebf4cd2eb5bcece57425704e62d35cc/docs/Server.md#setnotfoundhandler) will also need to have `beforeHandler` changed to `preHandler`.

**Comment by delvedor** [comment]
> Is this semver-major, or are you targeting next for convenience?

@mcollina for convenience, since we are already using the new `request` and `reply` objects.

File: fastify.js (hunk 4 above)
Comment (delvedor) [reply fastify.js]: It is not :D

**Comment by cemremengu** [comment]
So ideally checking for authentication should be done in `preValidation` right? 

I think it maybe confusing for people who consider authentication as a kind validation (as in validate the user)

File: lib/handleRequest.js (hunk 3 above)
Comment (delvedor) [reply lib/handleRequest.js]: Ok, I'll update also the rest of the codebase.

**Comment by delvedor** [comment]
> So ideally checking for authentication should be done in preValidation right?

Yes.

> I think it maybe confusing for people who consider authentication as a kind validation (as in validate the user)

I agree that's why I think we should discuss a little bit more about the hook name.

File: fastify.js [hunk 6]
```diff
@@ -585,13 +595,23 @@ function build (options) {
...
-          opts.beforeHandler.forEach((h, i) => {
-            opts.beforeHandler[i] = h.bind(_fastify)
-          })
+      if (opts.preHandler == null && opts.beforeHandler != null) {
+        beforeHandlerWarning()
+        opts.preHandler = opts.beforeHandler
+      }
+      if (opts.preHandler) {
+        if (Array.isArray(opts.preHandler)) {
+          opts.preHandler.map(hook => hook.bind(_fastify))
```
Comment (nwoltman) [inline fastify.js]: This should be:

```js
opts.preHandler = opts.preHandler.map(hook => hook.bind(_fastify))
```

It looks like there aren't any tests to test that `this` is the fastify instance inside `preHandler/preValidation` hooks that are passed to `.route()` as an array. It would be a good idea to add a test to catch this case.

Comment (nwoltman) [reply fastify.js]: That test wasn't failing because `preHandler` was set as a function. If it were an array of functions it would have failed.

**Comment by delvedor** [comment]
@mcollina completely agree!
A little recap for everybody: we should add an alias to `preValidation` to help users figure out which is the best hook to use for the authentication.
Which name do you prefer?

- `auth` 👍 
- `onAuth` 😄 
- `authentication` 🎉 
- `onAuthentication` ❤️ 

If you have other proposals, feel free to update this message (and drop a comment) :)

**Comment by delvedor** [comment]
@fastify/fastify I would love to have the opinion of all of you! :)

File: docs/Lifecycle.md [hunk 7]
```diff
@@ -14,13 +14,13 @@ Incoming Request
...
+                              │
+                    4**/5** ◀─┴─▶ Parsing
                                    │
-                             400 ◀─┴─▶ preHandler Hook
+                             415 ◀─┴─▶ Validation
                                          │
-                               4**/5** ◀─┴─▶ beforeHandler
+                                   400 ◀─┴─▶ preHandler Hook
                                                │
                                      4**/5** ◀─┴─▶ User Handler
```
Comment (allevo) [inline docs/Lifecycle.md]: Can we point out which phases are implemented by hooks?

File: fastify.js [hunk 8]
```diff
@@ -350,7 +350,7 @@ function build (options) {
     if (cb === undefined) return listenPromise(port, address, backlog)
 
     fastify.ready(function (err) {
-      if (err) return cb(err)
+      if (err != null) return cb(err)
```
Comment (allevo) [inline fastify.js]: Why `!=`?
`!== undefined || != null`?

Comment (delvedor) [reply fastify.js]: They are equivalent.

**Comment by delvedor** [comment]
> Are the benchmarks changed?

@allevo nope, as written at the beginning of the page :P

File: docs/Lifecycle.md (hunk 7 above)
Comment (delvedor) [reply docs/Lifecycle.md]: Can you rephrase the question?

File: fastify.js (hunk 8 above)
Comment (allevo) [reply fastify.js]: from performance point of view too?

Comment (delvedor) [reply fastify.js]: Yes.

Comment (StarpTech) [reply fastify.js]: Is it possible to check the performance of those tweaks in a reliable fashion?

Comment (delvedor) [reply fastify.js]: In this case is not a tweak. Is just "less characters for the same result" :P

Comment (StarpTech) [reply fastify.js]: That was a serious question 😄 What tools are you using except autocannon?

Comment (delvedor) [reply fastify.js]: Oops! Sorry!
`console.time`, [`nanobench`](https://github.com/mafintosh/nanobench) and [`fastbench`](https://github.com/mcollina/fastbench) :D

Comment (StarpTech) [reply fastify.js]: Thanks, 👍 same arsenal 😄

Comment (StarpTech) [reply fastify.js]: @delvedor they are all time-based. I always looking for ops-per-second measurment but I think this is not provided by V8 right?

File: docs/Lifecycle.md (hunk 7 above)
Comment (allevo) [reply docs/Lifecycle.md]: In that graph, some steps are done by the framework itself, other ones are hooks. Can we use a sign to point the hooks out? for example `*`

**Comment by cemremengu** [comment]
@delvedor make it `onAuth` then as the second most popular choice :)

**Comment by nwoltman** [comment]
The reason I voted for `auth` as the name instead of `onAuth` is because `onAuth` makes it sound like the user is hooking into Fastify's built-in authentication. However, Fastify doesn't have authentication built in, so this is misleading. The name `auth` makes it clear that the user is adding their own authentication and not hooking into an existing authentication process.

**Comment by delvedor** [comment]
@nwoltman I agree, but we can say the same also for the hooks, there no authentication support in Fastify so we are hooking what?
I think we should discuss more about this, maybe update `fastify-auth` and add there an API to handle the authentication, which internally will use the `preValidation` hook.

Something like:
```js
const fastify = require('fastify')
fastify.register(require('fastify-auth'))
...
fastify.addAuthentication(function myAuth (req, reply, next) { ... })
fastify.route({ ... })
```

**Comment by mcollina** [comment]
@delvedor That would actually be very nice.

**Comment by cemremengu** [comment]
@delvedor How would you `addAuthentication` to individual routes that way?

**Comment by delvedor** [comment]
@cemremengu that's a good question. It all depends on how we design the API in `fastify-hook`.

**Comment by cemremengu** [comment]
Maybe hooks for routes can somehow be redesigned to be functional (and chainable?) like `fastify.addHook` so we can do 

`routeA().routeB().addHook('preValidation', function(routes) => { /*auth*/ })`

this way we will be able to name our hooks to be like `addAuthentication`

**Comment by delvedor** [comment]
@cemremengu it's not possible, every API returns an instance of Fastify itself.
Probably the best option is to figure out a decorator or even better play with the `onRoute` hook.
```js
fastify.addHook('onRoute', opts => {
  if (opts.auth) {
    opts.preValidation = opts.auth
  }
})
```

I will prepare a pr for the `fastify-auth` plugin :)
